src/prefect/cli/deployment.py

```diff
@@ -273,12 +273,16 @@ async def apply(
             exit_with_error(f"'{path!s}' did not conform to deployment spec: {exc!r}")
 
         if upload:
-            file_count = await deployment.upload_to_storage()
-            if file_count:
-                app.console.print(
-                    f"Successfully uploaded {file_count} files to {deployment.location}",
-                    style="green",
-                )
+            if (
+                deployment.storage
+                and "put-directory" in deployment.storage.get_block_capabilities()
+            ):
+                file_count = await deployment.upload_to_storage()
+                if file_count:
+                    app.console.print(
+                        f"Successfully uploaded {file_count} files to {deployment.location}",
+                        style="green",
+                    )
```

Comment:
Should we add an `else` here that catches the case where a user is using a read-only storage block, but they've said to upload? Without it this seems like a silent failure.

Reply:
good call, updated